adoc/admin-gpus.adoc

```diff
@@ -0,0 +1,163 @@
+
+[source,bash]
+----
+# zypper install nvidia-container-toolkit
+----
+
+=== Test a GPU container image
+
+At this point, you should be able to run a container image that requires a GPU and directly access the device from the running container, for example using podman:
+
+[source,bash]
+----
+# podman run docker.io/nvidia/cuda nvidia-smi
+----
+
+=== Troubleshooting
+
+At this point, you should be able to run a container image using a GPU. If that is not working, check the following:
+
+Ensure your GPU is visible from the host system:
+
+[source,bash]
+----
+# lspci | grep -i nvidia
+# nvidia-smi
+----
+
+Ensure the kernel modules are loaded:
+
+[source,bash]
+----
+# lsmod | grep nvidia
+----
+
+If they are not, try loading them explicitly and check dmesg for an error indicating why they are missing:
+
+[source,bash]
+----
+# nvidia-modprobe
+# dmesg | tail
+----
+
+== Configure Kubernetes
+
+=== Install the device plugin
+
+The Kubernetes device plugin framework allows the kubelet to advertise system hardware resources that the Kubernetes scheduler can then use as hints to schedule workloads that require such devices. The Kubernetes device plugin from NVIDIA allows the kubelet to advertise NVIDIA GPUs it finds present on the worker node. Install the device plugin using kubectl:
+
+[source,bash]
+----
+$ kubectl create -f https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/1.0.0-beta6/nvidia-device-plugin.yml
+----
+
+=== Taint GPU workers
+
+In a heterogeneous cluster, it may be preferable to prevent scheduling pods that do not require a GPU on nodes with a GPU in order to ensure that GPU workloads are not competing for time on the hardware they need to run. To accomplish this, add a taint to the worker nodes that have GPUs:
+
+[source,bash]
+----
+$ kubectl taint nodes worker0 nvidia.com/gpu=:NoSchedule
+----
+
+or
+
+[source,bash]
+----
+$ kubectl taint nodes worker0 nvidia.com/gpu=:PreferNoSchedule
+----
+
+On every master node, add the ExtendedResourceToleration admission controller:
+
+[source,bash]
+----
+# sed -i -e '/--enable-admission-plugins=/ s/$/,ExtendedResourceToleration/' /etc/kubernetes/manifests/kube-apiserver.yaml
+----
+
+Wait for the kube-apiserver pod to be restarted. The ExtendedResourceToleration admission controller will automatically add the nvidia.com/gpu toleration to pods that request the nvidia.com/gpu resource.
+
+=== Test a GPU workload
+
+To test your installation you can create a pod that requests GPU devices:
+
+[source,bash]
+----
+$ cat <<EOF | kubectl apply -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  name: gpu-pod
+spec:
+  containers:
+    - name: cuda-container
+      image: nvidia/cuda:9.0-devel
+      resources:
+        limits:
+          nvidia.com/gpu: 1 # requesting 1 GPU
+    - name: digits-container
+      image: nvidia/digits:6.0
+      resources:
+        limits:
+          nvidia.com/gpu: 1 # requesting 1 GPU
+EOF
+----
+
```

Comment:
Do we need to have toleration since we taints GPU nodes?

Reply:
See the note on line 103:

> The ExtendedResourceToleration admission controller will automatically add the nvidia.com/gpu toleration to pods that request the nvidia.com/gpu resource.

I will change this to highlight this more prominently.

Reply:
Added on line 98.